# Tugas hari 4 pekan 2
by Arif Laksito<br />
arif.laksito@amikom.ac.id

<b>Note</b>: Pada kuis kali ini kita akan menggunakan data diamonds.csv yang bisa kalian download [disini](https://drive.google.com/uc?export=download&id=1qQiM9utQUThTPq8jk8stPMh_y2zfqYOd). setelah di download, kalian pindahkan data tersebut satu level dengan direktori file notebook ini.


In [2]:
import pandas as pd
df = pd.read_csv('diamonds-3.csv')
df.drop('Unnamed: 0', axis=1, inplace=True)

In [3]:
df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

Columns Metadata:

carat: berat diamond

cut: kualitas diamond

color: warna diamond

clarity: level kemurnian diamonds

depth: tinggi dari diamond tabel

table: lebar  dari diamond tabel

price: harga diamond ($US)

x: panjang

y: lebar

z: kedalaaman

<h2>Soal 1: Process Grouping</h2>

Jelaskan prosess yang terjadi dalam grouping data menggungkan method <b>groupby()</b> dari pandas.

jawab: Grouping adalah suatu teknik memisahkan data berdasarkan kriteria tertentu. Ketika melakukan grouping ada beberapa proses yang terjadi secara berurutan, yaitu :
- Splitting: Memisahkan data kedalam suatu group berdasarkan kriteria tertentu.
- Applying: Melakukan suatu operasi terhadap sekumpulan data di group-group tersebut.
- Combining: Menggabungkan data menjadi suatu struktur baru

<h2>Soal 2: Grouping All Column</h2>
    
Group data diamond berdasarkan cut (kualitas), lakukan aggregasi dengan fungsi rata2.

Kemudian jawab pertanyaan berikut.
- Berapa harga rata2 dari diamond berkualitas Premium

Expected Output:

![alt text](https://drive.google.com/uc?id=1GKanuLzNv_pYBFVRBmQywxQw39uQBvSl)

In [6]:
# code here
df.groupby('cut').mean()

carat      depth      table        price         x         y  \
cut                                                                          
Fair       1.046137  64.041677  59.053789  4358.757764  6.246894  6.182652   
Good       0.849185  62.365879  58.694639  3928.864452  5.838785  5.850744   
Ideal      0.702837  61.709401  55.951668  3457.541970  5.507451  5.520080   
Premium    0.891955  61.264673  58.746095  4584.257704  5.973887  5.944879   
Very Good  0.806381  61.818275  57.956150  3981.759891  5.740696  5.770026   

                  z  
cut                  
Fair       3.982770  
Good       3.639507  
Ideal      3.401448  
Premium    3.647124  
Very Good  3.559801

jawab: harga rata2 dari diamond berkualitas Premium adalah <strong>4584.257704</strong>

<h2>Soal 3: Grouping Some Column base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut (kualitas) dan color (warna) secara berurutan, dengan aggregasi pada column carat, table, dan price saja. dengan fungsi agregasi rata-rata.

Kemudian jawab pertanyaan berikut:
- Berapa harga diamond yang berkriteria premium dan berwarna J
- Berapa nilai dari carat diamond yang berkriteria very good dan berwarna H

Expected Output:

Note: gambar hanya sebagaian dari keseluruhan data

![alt text](https://drive.google.com/uc?id=1XB_uHor7ZCPQUH9uQjSfd6aC6_yJnjh9)

In [7]:
# code here
df.groupby(['cut','color'])[['carat','table','price']].mean()

carat      table        price
cut       color                                  
Fair      D      0.920123  58.969325  4291.061350
          E      0.856607  59.364732  3682.312500
          F      0.904712  59.453205  3827.003205
          G      1.023822  58.773248  4239.254777
          H      1.219175  58.696370  5135.683168
          I      1.198057  59.237143  4685.445714
          J      1.341176  58.917647  4975.655462
Good      D      0.744517  58.541541  3405.382175
          E      0.745134  58.779957  3423.644159
          F      0.775930  58.910891  3495.750275
          G      0.850896  58.471986  4123.482204
          H      0.914729  58.611111  4276.254986
          I      1.057222  58.773946  5078.532567
          J      1.099544  58.813029  4574.172638
Ideal     D      0.565766  55.965632  2629.094566
          E      0.578401  55.967461  2597.550090
          F      0.655829  55.924203  3374.939362
          G      0.700715  55.902375  3720.706388
          H      0.799525  55.965843  3889.334831
          I      0.913029  56.021357  4451.970377
          J      1.063594  56.012612  4918.186384
Premium   D      0.721547  58.718964  3631.292576
          E      0.717745  58.779461  3538.914420
          F      0.827036  58.679279  4324.890176
          G      0.841488  58.702360  4500.742134
          H      1.016449  58.792034  5216.706780
          I      1.144937  58.771849  5946.180672
          J      1.293094  58.874752  6294.591584
Very Good D      0.696424  58.041309  3470.467284
          E      0.676317  58.038875  3214.652083
          F      0.740961  57.848429  3778.820240
          G      0.766799  57.784428  3872.753806
          H      0.915948  57.903015  4535.390351
          I      1.046952  58.105150  5255.879568
          J      1.133215  58.277729  5103.513274

jawab:
- Harga diamond yang berkriteria premium dan berwarna J adalah <strong>6294.591584</strong>
- Nilai dari carat diamond yang berkriteria very good dan berwarna H adalah <strong>0.915948</strong>

<h2>Soal 4: Group By Multiple Aggregate Base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut(kualitas) dan color (warna) secara berurutan kemudian aggregasi objek group by nya dengan fungsi rata2 dan median.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai median dari harga diamond berkualitas good dan berwarna F

Expected Output:

Note : gambar hanya sebagaian dari keseluruhan data

![alt text](https://drive.google.com/uc?id=1fUPIse5ELrzj26ZyWBqsAVOUb6yf7UoD)

In [8]:
# code here
df.groupby(['cut','color']).agg(['mean','median'])

carat             depth             table         \
                     mean median       mean median       mean median   
cut       color                                                        
Fair      D      0.920123  0.900  64.048466   64.8  58.969325   58.0   
          E      0.856607  0.900  63.319643   64.8  59.364732   59.0   
          F      0.904712  0.900  63.508013   64.8  59.453205   59.0   
          G      1.023822  0.980  64.339809   65.0  58.773248   58.0   
          H      1.219175  1.010  64.585149   65.1  58.696370   58.0   
          I      1.198057  1.010  64.220571   65.1  59.237143   58.0   
          J      1.341176  1.030  64.357143   65.0  58.917647   58.0   
Good      D      0.744517  0.700  62.365710   63.4  58.541541   58.0   
          E      0.745134  0.700  62.203751   63.3  58.779957   58.0   
          F      0.775930  0.710  62.202310   63.2  58.910891   59.0   
          G      0.850896  0.900  62.526636   63.4  58.471986   58.0   
          H      0.914729  0.900  62.499573   63.4  58.611111   58.0   
          I      1.057222  1.000  62.475096   63.5  58.773946   58.0   
          J      1.099544  1.020  62.395765   63.4  58.813029   58.0   
Ideal     D      0.565766  0.500  61.678370   61.8  55.965632   56.0   
          E      0.578401  0.500  61.686677   61.8  55.967461   56.0   
          F      0.655829  0.530  61.675562   61.8  55.924203   56.0   
          G      0.700715  0.540  61.700164   61.8  55.902375   56.0   
          H      0.799525  0.700  61.733194   61.8  55.965843   56.0   
          I      0.913029  0.740  61.793598   61.9  56.021357   56.0   
          J      1.063594  1.030  61.821987   61.9  56.012612   56.0   
Premium   D      0.721547  0.580  61.169245   61.3  58.718964   59.0   
          E      0.717745  0.580  61.175909   61.3  58.779461   59.0   
          F      0.827036  0.760  61.259974   61.5  58.679279   59.0   
          G      0.841488  0.755  61.279480   61.4  58.702360   59.0   
          H      1.016449  1.010  61.322034   61.5  58.792034   59.0   
          I      1.144937  1.140  61.328641   61.6  58.771849   59.0   
          J      1.293094  1.250  61.390099   61.6  58.874752   59.0   
Very Good D      0.696424  0.610  61.750165   62.2  58.041309   58.0   
          E      0.676317  0.570  61.730250   62.0  58.038875   58.0   
          F      0.740961  0.700  61.721627   62.0  57.848429   58.0   
          G      0.766799  0.700  61.841279   62.1  57.784428   58.0   
          H      0.915948  0.900  61.968366   62.3  57.903015   58.0   
          I      1.046952  1.005  61.934551   62.2  58.105150   58.0   
          J      1.133215  1.060  61.902065   62.2  58.277729   58.0   

                       price                 x                y         \
                        mean  median      mean median      mean median   
cut       color                                                          
Fair      D      4291.061350  3730.0  6.018344  6.080  5.963190  6.010   
          E      3682.312500  2956.0  5.909063  6.070  5.858214  6.020   
          F      3827.003205  3035.0  5.990513  6.060  5.931122  5.990   
          G      4239.254777  3057.0  6.173822  6.125  6.114076  6.040   
          H      5135.683168  3816.0  6.579373  6.330  6.497393  6.250   
          I      4685.445714  3246.0  6.564457  6.340  6.493486  6.270   
          J      4975.655462  3302.0  6.747311  6.490  6.675882  6.460   
Good      D      3405.382175  2728.5  5.620076  5.690  5.633897  5.730   
          E      3423.644159  2420.0  5.617889  5.680  5.632454  5.710   
          F      3495.750275  2647.0  5.693443  5.760  5.709659  5.790   
          G      4123.482204  3340.0  5.850264  6.050  5.862824  6.070   
          H      4276.254986  3468.5  5.968789  6.110  5.974245  6.130   
          I      5078.532567  3639.5  6.253544  6.275  6.260536  6.260   
          J      4574.172638  3733.0  6.377003  6.440  6.386026  6.460   
Ideal     D      2629.094

jawab: Nilai median dari harga diamond berkualitas good dan berwarna F adalah <strong>59.0</strong>

<h2>Soal 5: Group By With Different Aggregation for specific Column</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian aggreagate dengan fungsi max pada column price dan fungsi min pada column carat.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai max dari price untuk diamond berkualitas Ideal

Expected Output:

![alt text](https://drive.google.com/uc?id=1B_z5PBkLe4hNJJJV3JgDGP9G--llytKf)

In [9]:
# code here
df.groupby('cut').agg({'price':'max','carat':'min'})

price  carat
cut                    
Fair       18574   0.22
Good       18788   0.23
Ideal      18806   0.20
Premium    18823   0.20
Very Good  18818   0.20

jawab: Nilai max dari price untuk diamond berkualitas Ideal adalah <strong>18806</strong>

<h2>Soal 6: Group By With Custom Function</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian lakukan aggregasi dengan fungsi standar deviasi pada column carat dan fungsi rentang nilai antara max dan min pada column price.

Kemudian jawab pertanyaan berikut:
- Berapa selisih harga tertinggi dan terendah dari diamond berkualitas Premium

Expected Output:

![alt text](https://drive.google.com/uc?id=1bl6_8KryXCxLfzpAf4YF80Qvt15a3uQ6)

In [11]:
# code here
def max_min(x):
    return x.max()-x.min()

df.groupby('cut').agg({'price':max_min,'carat':'std'})

price     carat
cut                       
Fair       18237  0.516404
Good       18461  0.454054
Ideal      18480  0.432876
Premium    18497  0.515262
Very Good  18482  0.459435

jawab: Selisih harga tertinggi dan terendah dari diamond berkualitas Premium adalah <strong>18497</strong>